Automattic/woocommerce-payments
# Prevent capturing payment for refunded orders

Fixes #2846

This change prevents an in-person payment to be captured for a partially or fully refunded order by throwing a 400 error from the API endpoint. We want to block such payments to avoid overcharging customers for unpaid orders, as can be seen in #2846.

##### Aside

So, what does a merchant do if they want to receive payment on a partially refunded order via cash on delivery method? As suggested in the linked issue, they should not refund an unpaid order in the first place. An order placed with cash on delivery (in-person) as the payment method would usually set its status to 'Processing'. A 'processing' order can be edited but only via the refund route. To edit an order (eg. remove an item) without issuing a refund, first change the order status from 'Processing' to 'Pending payment', and then perform the edits.

##### TODO

It may be useful to articulate and display the above messaging to merchants to avoid confusion.

#### Testing instructions

As this change affects a REST endpoint, we'll need to test the endpoint using a REST client such as Postman. The following instructions are similar to the steps to reproduce mentioned in the linked issue #2846.

1. Create a new order with two non-virtual/downloadable items, eg. a t-shirt ($10) and a hoodie ($45).
2. In wp-admin, ensure that the status of the new order is 'Processing'. Sometimes it may be 'On hold'; if that's the case for you, manually change it to 'Processing'.
3. Edit the order by refunding one item (say, the hoodie). The revised order total value should now be reduced by refunded item's value (eg. $45).
4. Create a new payment intent with an amount equal to the revised order value. Use Stripe Dashboard to create the intent - https://dashboard.stripe.com/test/connect/accounts/acct_<your_test_account_id>/payments/new.
5. Make a REST call* to the `capture_terminal_payment` endpoint, as shown in the screenshot. You should get an error 400 response. Before this change, it was error 500.

\* Tip: If you're trying to test WCPay Client REST endpoints in Postman and getting the error `Sorry, you are not allowed to do that.`, you may not have the correct auth set up: PbTz5e-12J-p2

<img width="1461" alt="image" src="https://user-images.githubusercontent.com/1288616/136206925-7f0b2938-d5ba-419b-87aa-1ed1b2b4ed14.png">

-------------------


**Post merge**



- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (does not apply)
